datafusion/src/physical_plan/math_expressions.rs

```diff
@@ -102,6 +103,31 @@ math_unary_function!("ln", ln);
 math_unary_function!("log2", log2);
 math_unary_function!("log10", log10);
 
+/// factorial SQL function
+pub fn factorial(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    match &args[0] {
+        ColumnarValue::Array(array) => {
+            let x1 = array.as_any().downcast_ref::<Float64Array>();
```

Comment:
I think by this point, the DataFusion coercion logic should have kicked in and you shouldn't have to handle the cases where array is not a `Float` array -- again I think following `sqrt` or other similar function might be helpful